File: 0006-zigzag-conversion/0006-zigzag-conversion.rb
def convert(s, numRows)
    return s if numRows == 1 || numRows >= s.length
    rows = Array.new(numRows) { "" }
    idx, step = 0, 1
    s.each_char do |c|
        rows[idx] += c
        if idx == 0
            step = 1
        elsif idx == numRows - 1
            step = -1
        end
        idx += step
    end
    rows.join("")
end

File: 0006-zigzag-conversion/0006-zigzag-conversion.js
function convert(s, numRows) {
    if (numRows === 1 || numRows >= s.length) {
        return s;
    }

    let rows = Array.from({ length: numRows }, () => '');
    let idx = 0, step = 1;

    for (let c of s) {
        rows[idx] += c;
        if (idx === 0) {
            step = 1;
        } else if (idx === numRows - 1) {
            step = -1;
        }
        idx += step;
    }

    return rows.join('');
}